### ISSUE TITLE
unload onStop restarts player

### ISSUE THREAD

**Issue by phloxic** [flowplayer/flash#61]
Yet another, yet different incarnation of an unload problem. Unloading onStop restarts the player - when it is not in fullscreen mode.

http://flowplayer.blacktrash.org/test/unload.html

``` js
clip: {
  onStop: function () {
    this.unload();
  }
}
```

Confirmed for Safari on MacOS > 10.6 - but might happen in other browsers too. @bbbo - perhaps you find time to test on windows.

Workaround:

``` js
clip: {
  onStop: function () {
    var player = this;
    setTimeout(function () {
      this.unload();
    }, 0);
  }
}
```

What does NOT help in this case is to apply https://github.com/flowplayer/flash/blob/master/core/src/javascript/flowplayer.js/flowplayer-src.js#L599 unconditionally.

I wonder whether https://github.com/flowplayer/flash/blob/master/core/src/javascript/flowplayer.js/flowplayer-src.js#L707 gets in the way/is triggered somehow.

**Comment by danrossi** [comment]
in safari it is unloading ok for me.

**Comment by danrossi** [comment]
window.onload = function () {
    $f("player1", "/flowplayer-3.2.15.swf", {
      clip: {
        url: "mp4:ccc_trailer2",
        provider: "rtmp",
        scaling: "fit",
        onStop: function () {
          this.unload();
        },
        onFinish: function () {
          this.unload();
        }
      },
      plugins: {
        controls: {stop: true},
        rtmp: {
          url: "flowplayer.rtmp-3.2.11.swf",
          netConnectionUrl: "rtmp://stream.blacktrash.org/cfx/st"
        }
      }
    });
  }; 

its something to do with this code in the header, works ok in the body.

**Comment by danrossi** [comment]
window.onload = function() {
        $f("player", "../build/flowplayer-3.2.15.swf",
                {
                    log: { level: 'debug', filter: 'org.flowplayer.controller._, org.flowplayer.rtmp._, org.flowplayer.menu._, org.flowplayer.bitrateselect._' },

```
                clip: {
                    scaling: "fit",
                    onStop: function () {
                        this.unload();
                    },
                    onFinish: function () {
                        this.unload();
                    },
                    url: "mp4:ccc_trailer2",


                    provider: 'rtmp'
                },
                plugins: {
                    controls: {stop: true},

                    rtmp: {
                        url: 'flowplayer.rtmp.swf',
                        netConnectionUrl: "rtmp://stream.blacktrash.org/cfx/st"

                    }
                }
            });
```

  }
      /*
  window.onload = function () {
      $f("player", "../build/flowplayer-3.2.15.swf", {
          clip: {
              url: "mp4:ccc_trailer2",
              provider: "rtmp",
              scaling: "fit",
              onStop: function () {
                  this.unload();
              },
              onFinish: function () {
                  this.unload();
              }
          },
          plugins: {
              controls: {stop: true},
              rtmp: {
                  url: "flowplayer.rtmp.swf",
                  netConnectionUrl: "rtmp://stream.blacktrash.org/cfx/st"
              }
          }
      });
  }; */

top one works bottom one does not exact same code or is it not ?

**Comment by phloxic** [comment]
Perhaps the logging leads to that tiny delay which is needed?

**Comment by danrossi** [comment]
having some issues with other problems here i have to resolve. when stopping it wont restart properly regardless of unload

**Comment by phloxic** [comment]
@danrossi - long shot, can you test here
http://flowplayer.blacktrash.org/test/unload.html
whether the following api hack would work (I don't have a problematic browser available):

```
diff --git a/core/src/javascript/flowplayer.js/flowplayer-src.js b/core/src/javascript/flowplayer.js/flowplayer-src.js
index 163c98e..b5b951f 100644
--- a/core/src/javascript/flowplayer.js/flowplayer-src.js
+++ b/core/src/javascript/flowplayer.js/flowplayer-src.js
@@ -187,6 +187,8 @@
                        }

                        self[evt] = function(fn) {
+                               if (evt === "onStop")
+                                       fn = setTimeout(fn, 0);
                                bind(listeners, evt, fn);
                                return self;
                        };
```

**Comment by danrossi** [comment]
works fine to me

**Comment by phloxic** [comment]
@danrossi - should be tested in Safari > 5, preferrably latest, on Mac and Windows - @bbbo, @anssip:
http://flowplayer.blacktrash.org/test/unload.html
Testing makes only sense if the current version exhibits the problematic behaviour, it's available at:
http://flowplayer.blacktrash.org/test/unload-bug.html
I can't confirm, with what I've here both versions work.

**Comment by bbbo** [comment]
In Safari 5.1.7 Win both examples simply return to the intiial state for me.

**Comment by phloxic** [comment]
@bbbo - yeah, I only observed it on a newer Mac with Safari > 5.x (don't remember which version).

**Comment by danrossi** [comment]
its not working in safari no

**Comment by phloxic** [comment]
@danrossi, thanks for testing - I take that a fix is needed then.
I have another, less invasive alternative which only narrows down the condition for the timeout in unload instead of adding another timeout. Could you test http://flowplayer.blacktrash.org/test/unload.html again?
And please also test stop when in fullscreen - that's important.
Thanks.

**Comment by anssip** [comment]
Tested on latest Safari: Works when stopped in fullscreen. Does not work in non-fullscreen, the splash shows briefly but then the player loads on top of it and starts playing from the start.

**Comment by phloxic** [comment]
Right. Then we need the fix with the extra timeout. Thanks.

**Comment by phloxic** [comment]
@anssip - will submit patches shortly. Meanwhile you can test the fix at http://flowplayer.blacktrash.org/test/unload.html
Without fix: http://flowplayer.blacktrash.org/test/unload-bug.html - please test this one in Chrome, if possible I want to restrict the fix to Safari > 5.
Thanks.

**Comment by anssip** [comment]
Does not work for me in Safari (the 1st page). Reloads the player in non-fullscreen when stopping.

The page without fix works on Chrome.

**Comment by phloxic** [comment]
Did you empty the cache ;-)
Give it one more try. If it doesn't work, then I don't know how to implement the workaround I outlined in the original report in the api. The workaround is deployed at http://www.blacktrash.org/underdogma if you care to check.

**Comment by danrossi** [comment]
Is this still an issue ?

**Comment by phloxic** [comment]
Yes, just try http://flowplayer.blacktrash.org/test/unload-bug.html in Safari.
I haven't found a way to fix this nicely in the api script.

**Comment by danrossi** [comment]
ok thanks for that.
